lib/acl-checker.js

```diff
@@ -81,127 +83,66 @@ class ACLChecker {
+   * @param options {Object} Options hashmap
+   * @param [options.origin] Request's `Origin:` header
+   * @param [options.host] Request's host URI (with protocol)
+   */
   findRule (graph, user, mode, resource, accessType, acl, callback, options) {
     const debug = this.debug
     if (!graph || graph.length === 0) {
       debug('ACL ' + acl + ' is empty')
-      return callback(new Error('No policy found'))
+      return callback(new Error('No policy found - empty ACL'))
     }
-    debug('Found policies in ' + acl)
-    // Check for mode
-    var statements = this.getMode(graph, mode)
-    if (mode === 'Append') {
-      statements = statements.concat(this.getMode(graph, 'Write'))
+    let isContainer = accessType.startsWith('default')
+    let aclOptions = {
+      aclSuffix: this.suffix,
+      graph: graph,
+      host: options.host,
+      origin: options.origin,
+      rdf: rdf,
+      strictOrigin: this.strictOrigin,
+      isAcl: (uri) => { return this.isAcl(uri) },
+      aclUrlFor: (uri) => { return this.aclUrlFor(uri) }
     }
-    var self = this
-    async.some(
-      statements,
-      function (statement, done) {
-        var statementSubject = statement.subject.uri
-        // Check for origin
-        var matchOrigin = self.matchOrigin(graph, statementSubject,
-          options.origin, options.host)
-        if (!matchOrigin) {
-          debug('The request does not match the origin')
-          return done(false)
-        }
-        // Check for accessTo/defaultForNew
-        if (!self.isAcl(resource) || accessType === 'defaultForNew') {
-          debug('Checking for accessType:' + accessType)
-          var accesses = self.matchAccessType(graph, statementSubject, accessType,
-            resource)
-          if (!accesses) {
-            debug('Cannot find accessType ' + accessType)
-            return done(false)
-          }
-        }
-        // Check for Agent
-        var agentStatements = []
-        if (user) {
-          agentStatements = graph.match(
-            statementSubject,
-            'http://www.w3.org/ns/auth/acl#agent',
-            user)
-        }
-        if (agentStatements.length) {
-          debug(mode + ' access allowed (as agent) for: ' + user)
-          return done(true)
-        }
-        debug('Inspect agentClass')
-        // Check for AgentClass
-        return self.findAgentClass(graph, user, mode, resource, statementSubject,
-          done)
-      },
-      function (found) {
-        if (!found) {
+    let acls = new PermissionSet(resource, acl, isContainer, aclOptions)
+    acls.checkAccess(resource, user, mode)
+      .then(hasAccess => {
+        if (hasAccess) {
+          debug(`${mode} access permitted to ${user}`)
+          return callback()
+        } else {
+          debug(`${mode} access not permitted to ${user}`)
           return callback(new Error('Acl found but policy not found'))
         }
-        return callback(null)
+      })
+      .catch(err => {
```

Comment:
Could you clarify the distinction between a rejected promise and a resolved promise with a falsy value in `checkAccess`?